dan_layer/consensus/src/hotstuff/on_receive_vote.rs

```diff
@@ -1,109 +1,142 @@
-        //
-        //     let votes: Vec<VoteMessage> = tx.get_received_votes_for(msg.local_node_hash())?;
-        //
-        //     if votes.len() == valid_committee.consensus_threshold() {
-        //         let validator_metadata = votes.iter().map(|v| v.validator_metadata().clone()).collect();
-        //         let proofs = votes
-        //             .iter()
-        //             .map(|v| v.merkle_proof().unwrap().clone())
-        //             .collect::<Vec<_>>();
-        //
-        //         let merged_proof = MergedBalancedBinaryMerkleProof::create_from_proofs(proofs).unwrap();
-        //         let leaf_hashes = votes.iter().map(|v| v.node_hash()).collect::<Vec<_>>();
-        //
-        //         // TODO: Check all votes
-        //         let main_vote = votes.get(0).unwrap();
-        //
-        //         let qc = QuorumCertificate::new(
-        //             node.payload_id(),
-        //             node.payload_height(),
-        //             *node.hash(),
-        //             node.height(),
-        //             node.shard(),
-        //             node.epoch(),
-        //             self.public_key.clone(),
-        //             main_vote.decision(),
-        //             main_vote.all_shard_pledges().clone(),
-        //             validator_metadata,
-        //             Some(merged_proof),
-        //             leaf_hashes,
-        //         );
-        //         self.update_high_qc(&mut tx, node.proposed_by().clone(), qc)?;
-        //
-        //         on_propose = Some((node.shard(), node.payload_id()));
-        //     }
-        //
-        //     // commit the transaction
-        //     tx.commit()?;
-        // }
-        //
-        // // Propose the next node
-        // if let Some((shard_id, payload_id)) = on_propose {
-        //     // TODO: This should go in a some component that controls message flows and events
-        //     let epoch = self.epoch_manager.current_epoch().await?;
-        //     let committee = self.epoch_manager.get_committee(epoch, shard_id).await?;
-        //     if committee.is_empty() {
-        //         return Err(HotStuffError::NoCommitteeForShard { shard: shard_id, epoch });
-        //     }
-        //     if self.is_leader(payload_id, shard_id, &committee)? {
-        //         self.leader_on_propose(shard_id, payload_id).await?;
-        //     }
-        // }
-        // Ok(())
+
+        let addr = self.epoch_manager.get_our_validator_addr(message.epoch).await?;
+        let committee = self.epoch_manager.get_local_committee(message.epoch).await?;
+        if !committee.contains(&addr) {
+            return Err(HotStuffError::ReceivedMessageFromNonCommitteeMember {
+                epoch: message.epoch,
+                sender: from.to_string(),
+                context: "OnVoteReceived".to_string(),
+            });
+        }
+
+        if !self.leader_strategy.is_leader(&addr, &committee, &message.block_id, 0) {
```

Comment:
There are actually two leaders involved here. The leader that proposes a block, and then the leader who proposes the next block will be the one who receives the vote. Given that, it's not clear whether this logic is correct. It should maybe use a different method `is_leader_for_receiving_votes()` or `is_leader_for_next_block`

Reply:
The votes are sent to the leader for the block which was originally proposed by another leader. So here we check that we're the leader for the block being voted for. `leader_for_next_block(current) == is_leader(next)`